Building-provided self-healing customization

Starkku opened the pull request:
- It is now possible to set a global cap for the effects of `InfantryGainSelfHeal` and `UnitsGainSelfHeal` by setting `InfantryGainSelfHealCap` & `UnitsGainSelfHealCap` under `[General], respectively.
- It is also possible to change the pip frames displayed from `pips.shp` individually for infantry, units and buildings by setting the frames for infantry & unit self-healing on `Pips.SelfHeal.Infantry/Units/Buildings` under `[AudioVisual]`, respectively.
- Whether or not a TechnoType benefits from effects of `InfantryGainSelfHeal` or `UnitsGainSelfHeal` buildings or neither can now be controlled by setting `SelfHealGainType`.
  - If `SelfHealGainType` is not set, InfantryTypes and VehicleTypes with `Organic` set to true gain self-healing from `InfantryGainSelfHeal`, other VehicleTypes from `UnitsGainSelfHeal` and AircraftTypes & BuildingTypes never gain self-healing.

In `rulesmd.ini`:
```ini
[General]
InfantryGainSelfHealCap=      ; int, maximum amount of InfantryGainSelfHeal that can be in effect at once, must be 1 or higher
UnitsGainSelfHealCap=         ; int, maximum amount of UnitsGainSelfHeal that can be in effect at once, must be 1 or higher

[AudioVisual]
Pips.SelfHeal.Infantry=13,20  ; int, frames of pips.shp for infantry & unit-self healing pips, respectively
Pips.SelfHeal.Units=13,20     ; int, frames of pips.shp for infantry & unit-self healing pips, respectively
Pips.SelfHeal.Buildings=13,20 ; int, frames of pips.shp for infantry & unit-self healing pips, respectively

[SOMETECHNO]                  ; TechnoType
SelfHealGainType=             ; Self-Heal Gain Type Enumeration (none|infantry|units)
```

Additionally the pips now follow unit's `PixelSelectionBracketDelta` like the health bar pips do, and if self-healing is added for BuildingTypes the pips will adjust their position according to the building's `Height` like the health bar does, and buildings with `SelfHealing` will now correctly restore the undamaged graphics when healed above `[AudioVisual]` -> `ConditionYellow`.

Starkku commented:
Refactored the self-healing (not pip-drawing) related code and added an optional global cap for effects of `Infantry/UnitsGainSelfHeal`.

AlliedG commented:
**Terror Drone is an InfantryType with SelfHealGainType=units
Engineer has SelfHealGainType=none
Hover tank is SelfHealGainType=infantry without Organic=yes tag**

![image](https://user-images.githubusercontent.com/84816840/162594549-f0a8ea3d-fb94-4c4d-8c65-4a4f831b646e.png)

**Self healing on building types (also tested on con yard)**

![image](https://user-images.githubusercontent.com/84816840/162594787-9093bf34-f681-47f9-816e-810f737d01ad.png)
![image](https://user-images.githubusercontent.com/84816840/162595357-b3cd3928-33ae-4cdd-bec8-2a8f9751f103.png)


Visual bug with self healing structures. If the structure is damaged enough to show damage frames and it self heals back to green health, the non damaged frame will not display unless the structure attacks (it magically reverts to normal state) or if it undeploys and redeploys.

![image](https://user-images.githubusercontent.com/84816840/162595282-ceae7e2e-64ef-4795-bfde-68672a248820.png)

**Highlighted turret showing damaged frames after self healing from critical health**

![image](https://user-images.githubusercontent.com/84816840/162595521-8e8a3fc6-512e-4061-8042-1b197cb9a820.png)

**Non selected construction yard is full health no damage taken, selected construction yard is showing damaged frames after self healing from critical damage**

[AudioVisual] tags worked as expected

[General] tags worked as expected (tested by having structure provide InfantryGainSelfHeal=10
UnitsGainSelfHeal=10) and setting cap to 1.

Starkku commented:
- Fixed building self-heal pip position
- Fixed  building graphics not reverting to undamaged when healed over AudioVisual->ConditionYellow by self-healing, both SelfHealGain and regular SelfHealing

8abs commented:
How about building a feature that infantries/units/buildings' self-healing rate&amount could be given (or changed) by warhead.

AlliedG commented:
self-heal pip position adjusted as expected
damaged structure reverts to undamaged state when healed over AudioVisual -> ConditionYellow

![image](https://user-images.githubusercontent.com/84816840/167273557-94d67308-3d76-4ac7-ba95-a71e3bc3339b.png)

Metadorius commented on docs/Fixed-or-Improved-Logics.md:
```diff
@@ -141,6 +143,28 @@ Gravity=6.0             ; double
 
+### Building-provided self-healing customization
+
+- It is now possible to set a global cap for the effects of `InfantryGainSelfHeal` and `UnitsGainSelfHeal` by setting `InfantryGainSelfHealCap` & `UnitsGainSelfHealCap` under `[General], respectively.
```
```suggestion
- It is now possible to set a global cap for the effects of `InfantryGainSelfHeal` and `UnitsGainSelfHeal` by setting `InfantryGainSelfHealCap` & `UnitsGainSelfHealCap` under `[General]`, respectively.
```

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -517,6 +518,157 @@ void TechnoExt::UpdateSharedAmmo(TechnoClass* pThis)
+		if (pExt->SelfHealGainType.isset() && pExt->SelfHealGainType.Get() == SelfHealGainType::None)
+			return;
+
+		if (pThis->Owner->InfantrySelfHeal > 0 &&
+			((!pExt->SelfHealGainType.isset() && (pThis->WhatAmI() == AbstractType::Infantry ||
+				pThis->GetTechnoType()->Organic && pThis->WhatAmI() == AbstractType::Unit)) ||
+				pExt->SelfHealGainType.isset() && pExt->SelfHealGainType.Get() == SelfHealGainType::Infantry))
```
This condition is somewhat badly readable, perhaps you can split it by logical parts as a few local bool variables?

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -517,6 +518,157 @@ void TechnoExt::UpdateSharedAmmo(TechnoClass* pThis)
+			drawPip = true;
+			selfHealFrames = RulesClass::Instance->SelfHealInfantryFrames;
+		}
+		else if (pThis->Owner->UnitsSelfHeal > 0 &&
+			((!pExt->SelfHealGainType.isset() && pThis->WhatAmI() == AbstractType::Unit) ||
+				pExt->SelfHealGainType.isset() && pExt->SelfHealGainType.Get() == SelfHealGainType::Units))
```
Same here.

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -517,6 +518,157 @@ void TechnoExt::UpdateSharedAmmo(TechnoClass* pThis)
+		if (pThis->WhatAmI() == AbstractType::Unit || pThis->WhatAmI() == AbstractType::Aircraft)
+		{
+			pipFrames = RulesExt::Global()->Pips_SelfHeal_Units;
+			xOffset = 33;
+			yOffset = -32 + pThis->GetTechnoType()->PixelSelectionBracketDelta;
+		}
+		else if (pThis->WhatAmI() == AbstractType::Infantry)
+		{
+			pipFrames = RulesExt::Global()->Pips_SelfHeal_Infantry;
+			xOffset = 25;
+			yOffset = -35 + pThis->GetTechnoType()->PixelSelectionBracketDelta;
+		}
+		else
+		{
+			auto pType = abstract_cast<BuildingTypeClass*>(pThis->GetTechnoType());
+			int fHeight = pType->GetFoundationHeight(false);
+
+			pipFrames = RulesExt::Global()->Pips_SelfHeal_Buildings;
+			xOffset = 12 + 30 * fHeight;
+			yOffset = 15 + -15 * fHeight + pType->Height * -15;
```
Magic numbers?

Starkku replied:
It would make sense to define cell width / height constants as well as leptons per cell somewhere globally.

As for the other numbers, They are just defaults that were constants in the original game code. Could make them customizable but I don't think most people would use that.

Metadorius reviewed (changes requested):
Looks good apart from a few code readability issues. Not sure if magic numbers can be dealt with here, may be good to deal with them post-merge since IIRC there are a few such issues scattered across the codebase.

Starkku commented:
Fixed the issues, made the pixel offsets for pips customizable in case someone may wish to adjust the positions (default values are listed in documentation). Will look into fixing the other 'magic numbers' after merging.
